File: mutate4py/src/mutate4py/manifest.py
from __future__ import annotations

import json
from pathlib import Path

from mutate4py.model import (
    ChangedScopes,
    DifferentialManifest,
    MutationScope,
    SourceAnalysis,
)


class ManifestStore:
    def __init__(self, workspace_root: str) -> None:
        self.workspace_root = Path(workspace_root)

    def read(self, file_path: str) -> DifferentialManifest | None:
        path = self._manifest_path(file_path)
        if not path.exists():
            return None
        payload = json.loads(path.read_text(encoding="utf-8"))
        return DifferentialManifest(
            version=payload["version"],
            module_hash=payload["module_hash"],
            scopes=[MutationScope(**scope) for scope in payload["scopes"]],
        )

    def write(self, file_path: str, analysis: SourceAnalysis) -> None:
        path = self._manifest_path(file_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "version": 1,
            "module_hash": analysis.module_hash,
            "scopes": [scope.__dict__ for scope in analysis.scopes],
        }
        path.write_text(json.dumps(payload, indent=2), encoding="utf-8")

    def changed_scopes(self, file_path: str, analysis: SourceAnalysis) -> ChangedScopes:
        manifest = self.read(file_path)
        if manifest is None:
            return ChangedScopes(manifest_present=False, module_hash_changed=False)
        if manifest.module_hash == analysis.module_hash:
            return ChangedScopes(manifest_present=True, module_hash_changed=False)

        previous = {scope.id: scope.semantic_hash for scope in manifest.scopes}
        unregistered: set[str] = set()
        violations: set[str] = set()
        for scope in analysis.scopes:
            previous_hash = previous.get(scope.id)
            if previous_hash is None:
                unregistered.add(scope.id)
            elif previous_hash != scope.semantic_hash:
                violations.add(scope.id)
        return ChangedScopes(
            manifest_present=True,
            module_hash_changed=True,
            unregistered_scope_ids=unregistered,
            manifest_violation_scope_ids=violations,
        )

    def _manifest_path(self, file_path: str) -> Path:
        relative = Path(file_path)
        return (
            self.workspace_root
            / ".mutate"
            / "manifests"
            / relative.parent
            / f"{relative.name}.json"
        )

File: mutate4rs/src/manifest.rs
use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};

use crate::model::{ChangedScopes, DifferentialManifest, SourceAnalysis};

pub struct ManifestStore {
    workspace_root: PathBuf,
}

impl ManifestStore {
    pub fn new(workspace_root: PathBuf) -> Self {
        Self { workspace_root }
    }

    pub fn read(&self, file_arg: &str) -> Result<Option<DifferentialManifest>, String> {
        let path = self.manifest_path(file_arg);
        if !path.exists() {
            return Ok(None);
        }
        let content = fs::read_to_string(path).map_err(|err| err.to_string())?;
        serde_json::from_str(&content)
            .map(Some)
            .map_err(|err| err.to_string())
    }

    pub fn write(&self, file_arg: &str, analysis: &SourceAnalysis) -> Result<(), String> {
        let path = self.manifest_path(file_arg);
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).map_err(|err| err.to_string())?;
        }
        let manifest = DifferentialManifest {
            version: 1,
            module_hash: analysis.module_hash.clone(),
            scopes: analysis.scopes.clone(),
        };
        let content = serde_json::to_string_pretty(&manifest).map_err(|err| err.to_string())?;
        fs::write(path, content).map_err(|err| err.to_string())
    }

    pub fn changed_scopes(
        &self,
        file_arg: &str,
        analysis: &SourceAnalysis,
    ) -> Result<ChangedScopes, String> {
        let Some(manifest) = self.read(file_arg)? else {
            return Ok(ChangedScopes::default());
        };
        if manifest.module_hash == analysis.module_hash {
            return Ok(ChangedScopes {
                manifest_present: true,
                module_hash_changed: false,
                ..ChangedScopes::default()
            });
        }

        let previous = manifest
            .scopes
            .into_iter()
            .map(|scope| (scope.id, scope.semantic_hash))
            .collect::<BTreeMap<_, _>>();
        let mut changed = ChangedScopes {
            manifest_present: true,
            module_hash_changed: true,
            ..ChangedScopes::default()
        };
        for scope in &analysis.scopes {
            match previous.get(&scope.id) {
                None => {
                    changed.unregistered_scope_ids.insert(scope.id.clone());
                }
                Some(old_hash) if old_hash != &scope.semantic_hash => {
                    changed
                        .manifest_violation_scope_ids
                        .insert(scope.id.clone());
                }
                _ => {}
            }
        }
        Ok(changed)
    }

    fn manifest_path(&self, file_arg: &str) -> PathBuf {
        let relative = Path::new(file_arg);
        self.workspace_root
            .join(".mutate")
            .join("manifests")
            .join(relative.parent().unwrap_or_else(|| Path::new("")))
            .join(format!(
                "{}.json",
                relative
                    .file_name()
                    .and_then(|name| name.to_str())
                    .unwrap_or("unknown")
            ))
    }
}
